packages/tpetra/core/src/Tpetra_Details_residual.hpp

```diff
@@ -121,11 +195,102 @@ void localResidual(const CrsMatrix<SC,LO,GO,NO> &  A,
        std::runtime_error, "X, Y and R may not alias one another.");
   }
       
+#ifdef TPETRA_DETAILS_USE_REFERENCE_RESIDUAL
   // This is currently a "reference implementation" waiting until Kokkos Kernels provides
   // a residual kernel.
   SC one = Teuchos::ScalarTraits<SC>::one(), negone = -one, zero = Teuchos::ScalarTraits<SC>::zero();    
   A.localApply(X,R,Teuchos::NO_TRANS, one, zero);
   R.update(one,B,negone);
+#else
+  using execution_space = typename CrsMatrix<SC,LO,GO,NO>::execution_space;
+
+  if (A_lcl.numRows() == 0) {
+    return;
+  }
+
+  int team_size = -1;
+  int vector_length = -1;
+  int64_t rows_per_thread = -1;
+
+  const int64_t rows_per_team = 
+    residual_launch_parameters<execution_space>(A_lcl.numRows (), A_lcl.nnz (), rows_per_thread, team_size, vector_length);
+  int64_t worksets = (B_lcl.extent (0) + rows_per_team - 1) / rows_per_team;
+
+  using policy_type = typename Kokkos::TeamPolicy<execution_space>;
+  using team_member = typename policy_type::member_type;
+
+  using residual_value_type = typename  MultiVector<SC,LO,GO,NO>::dual_view_type::t_dev::non_const_value_type;
+  using KAT = Kokkos::ArithTraits<residual_value_type>;
+
+  policy_type policy (1, 1);
+  if (team_size < 0) {
+    policy = policy_type (worksets, Kokkos::AUTO, vector_length);
+  }
+  else {
+    policy = policy_type (worksets, team_size, vector_length);
+  }
+
+  bool is_vector = (X_lcl.extent(1) == 1);
+
+  if(is_vector) {
+    // Vector case
+    // Kernel interior shamelessly horked from Ifpack2_Details_ScaledDampedResidual_def.hpp
+    Kokkos::parallel_for("residual-vector",policy,KOKKOS_LAMBDA(const team_member& dev) {
+        // NOTE: It looks like I should be able to get this data up above, but if I try to
+        // we get internal compiler errors.  Who knew that gcc tried to "gimplify"?
```

Comment:
Sometimes the internal compiler errors go away if you capture inside by value, instead of by reference.  Also, always pass the loop index by (const) value, not by const reference.

Reply:
On the other hand, if you capture inside by value, you might have bigger kernels that take up more GPU registers etc.

Reply:
I know you're supposed to pass PODs by const value.  team_member should be treated the same way?

Reply:
`team_member` is a struct, so it's probably better to pass by const reference.